### PR TITLE
Updated system exit extension to expose last return code

### LINKED ISSUE #2156: System Exit Extension: Test fails though it should pass
**Kotest 4.4.2**

I encountered strange behaviour when using the `System Exit Extension`:

Minimum working example:

```
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.extensions.system.*
import io.kotest.matchers.shouldBe
import picocli.CommandLine
import picocli.CommandLine.Command
import java.util.concurrent.Callable
import kotlin.system.exitProcess

class SystemExitTest : StringSpec() {

    override fun listeners() = listOf(SpecSystemExitListener)

    init {
        "Test passes as expected" {
            shouldThrow<SystemExitException> {
                exitProcess(CommandLine(PicocliApp()).execute())
            }.exitCode shouldBe 42
        }

        "Test fails though it should pass" {
            shouldThrow<SystemExitException> {
                CommandLine(PicocliApp2()).execute()
            }.exitCode shouldBe 42
        }
    }
}

@Command
class PicocliApp : Callable<Int> {
    override fun call(): Int {
        return 42
    }
}

@Command
class PicocliApp2 : Callable<Int> {
    override fun call(): Int {
        exitProcess(42)
    }
}
```

I'm testing a command line application based on the [https://picocli.info](picocli) framework.

**First test:** PicocliApp app is called, returns an exit code, top level process exits with returned code. Everything is fine here.

**Second test:** PicocliApp app is called, app exits inside the overriden `call` function. Test fails though it should pass.

I tested the both same tests using Stefan Birkner's [System rules](https://stefanbirkner.github.io/system-rules/index.html). **Result:** both tests pass (as expected).

I admit that it is a bit unusual to exit from the `call` function of a command. Certainly the better way of coding is following the convention to return an exit code and call System.exit from the main method with the return value of the `execute` method. Nonetheless, test 2 contains valid code which should be handled correctly from the test framework.

Thanks for investigating!

### PR DESCRIPTION
**Pull request by sksamuel**
Closes #2156

### CODE REVIEW

File: kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemExitExtensions.kt [hunk 1]
```diff
@@ -80,7 +87,12 @@ class SystemExitException(val exitCode: Int) : RuntimeException()
 @Suppress("OverridingDeprecatedMember", "DEPRECATION")
 class NoExitSecurityManager(private val originalSecurityManager: SecurityManager?) : SecurityManager() {
 
-    override fun checkExit(status: Int) = throw SystemExitException(status)
+   var lastExitCode: Int = -1
```
Comment (LeoColman) [inline kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemExitExtensions.kt]: Perhaps changing this to be nullable instead of -1?

Comment (sksamuel) [reply kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemExitExtensions.kt]: Yes good call. Done.

File: kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemExitExtensions.kt [hunk 2]
```diff
@@ -50,21 +51,30 @@ object SystemExitListener : TestListener {
...
+      if (previousSecurityManagers.contains(spec::class.toDescription()))
+         System.setSecurityManager(previousSecurityManagers[spec::class.toDescription()])
+      else
+         System.setSecurityManager(null)
+   }
+
+   fun lastExitCode(): Int? = when (val manager = System.getSecurityManager()) {
+      is NoExitSecurityManager -> manager.lastExitCode
+      else -> null
+   }
```
Comment (LeoColman) [inline kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemExitExtensions.kt:73]:
```suggestion
   fun lastExitCode() = (System.getSecurityManager() as? NoExitSecurityManager)?.lastExitCode
```
IDK if better, but more concise

Comment (sksamuel) [reply kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemExitExtensions.kt:73]: Harder to read imo

File: kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemExitExtensions.kt [hunk 3]
```diff
@@ -50,21 +51,30 @@ object SystemExitListener : TestListener {
...
+
+   fun shouldHaveExitCode(code: Int) {
+      lastExitCode() shouldBe code
+   }
```
Comment (LeoColman) [inline kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemExitExtensions.kt:77]:
```suggestion
   fun shouldHaveExitCode(code: Int) {
      lastExitCode() shouldBe code
   }
   
   fun shouldNotHaveExitCode(code: Int) {
     lastExitCode() shouldNotBe code
   }
   
```

**Review by LeoColman — APPROVED** [review]

**Pull request merged**